File: packages/dara-core/cypress/integration/persistence.ts
import { loginBeforeRoute } from '../support/utils';

describe('Persistence API', () => {
    beforeEach(loginBeforeRoute('/persistence'));

    it('Persists variable with persist_value=True', () => {
        cy.cardContent('Simple scenario').within(() => {
            // Before changes
            cy.contains('div', 'PERSISTED_VALUE:').next().find('input').should('have.value', 'test');
            cy.contains('div', 'STATIC_VALUE:').next().find('input').should('have.value', 'static');

            // Update both inputs
            cy.get('input[value="test"]').clear().type('updated');
            cy.get('input[value="static"]').clear().type('updated');
        });

        // After reload, value with persist_value should be persisted
        cy.visit('/persistence').then(() => {
            cy.cardContent('Simple scenario').within(() => {
                cy.contains('div', 'PERSISTED_VALUE:').next().find('input').should('have.value', 'updated');
                cy.contains('div', 'STATIC_VALUE:').next().find('input').should('have.value', 'static');
            });
        });
    });

    it('Persistance works with nested variables', () => {
        cy.cardContent('Nested scenario').within(() => {
            // Before changes
            cy.contains('div', 'PERSISTED_VALUE:').next().find('input').should('have.value', 'test');

            // Update both inputs
            cy.get('input[value="test"]').clear().type('updated');

            // Wait for change to be persisted - wait because of debouncing
            cy.wait(1000);
        });

        // After reload, value with persist_value should be persisted
        cy.visit('/persistence').then(() => {
            cy.cardContent('Nested scenario').within(() => {
                cy.contains('div', 'PERSISTED_VALUE:').next().find('input').should('have.value', 'updated');
            });
        });
    });
});

File: packages/dara-core/cypress/dara/pages/persistence.py
from dara.components import Card, Input, Stack, Text
from dara.core.interactivity import Variable


def persistence():
    """
    Test Persistence API
    """
    # Basic scenario
    persisted_value = Variable('test', persist_value=True)
    static_value = Variable('static')

    simple_scenario = Stack(
        Text('PERSISTED_VALUE:'), Input(value=persisted_value), Text('STATIC_VALUE:'), Input(value=static_value)
    )

    # Nested scenario
    nested_persisted_value = Variable({'nested': {'value': 'test'}}, persist_value=True)
    nested_scenario = Stack(Text('PERSISTED_VALUE:'), Input(value=nested_persisted_value.get('nested').get('value')))

    return Stack(Card(simple_scenario, title='Simple scenario'), Card(nested_scenario, title='Nested scenario'))
